core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,618 @@
+struct PriorityIndex {
+    priority: u64,
+    index: usize,
+}
+
+/// Calculate priority for a transaction:
+///
+/// The priority is calculated as:
+/// P = R / (1 + C)
+/// where P is the priority, R is the reward,
+/// and C is the cost towards block-limits.
+///
+/// Current minimum costs are on the order of several hundred,
+/// so the denominator is effectively C, and the +1 is simply
+/// to avoid any division by zero due to a bug - these costs
+/// are estimate by the cost-model and are not direct
+/// from user input. They should never be zero.
+/// Any difference in the prioritization is negligible for
+/// the current transaction costs.
+fn calculate_priority(
+    transaction: &RuntimeTransaction<SanitizedTransactionView<&[u8]>>,
+    bank: &Bank,
+) -> Option<u64> {
+    let compute_budget_limits = transaction
+        .compute_budget_instruction_details()
+        .sanitize_and_convert_to_compute_budget_limits(&bank.feature_set)
+        .ok()?;
+    let fee_budget_limits = FeeBudgetLimits::from(compute_budget_limits);
+
+    // Manually estimate fee here since currently interface doesn't allow a on SVM type.
+    // Doesn't need to be 100% accurate so long as close and consistent.
+    let prioritization_fee = fee_budget_limits.prioritization_fee;
+    let signature_details = transaction.signature_details();
+    let signature_fee = signature_details
+        .total_signatures()
+        .saturating_mul(bank.fee_structure().lamports_per_signature);
+    let fee = signature_fee.saturating_add(prioritization_fee);
+
+    let cost = CostModel::estimate_cost(
+        transaction,
+        transaction.program_instructions_iter(),
+        transaction.num_requested_write_locks(),
+        &bank.feature_set,
+    );
+
+    // We need a multiplier here to avoid rounding down too aggressively.
+    // For many transactions, the cost will be greater than the fees in terms of raw lamports.
+    // For the purposes of calculating prioritization, we multiply the fees by a large number so that
+    // the cost is a small fraction.
+    // An offset of 1 is used in the denominator to explicitly avoid division by zero.
+    const MULTIPLIER: u64 = 1_000_000;
+    Some(
+        MULTIPLIER
+            .saturating_mul(fee)
+            .wrapping_div(cost.sum().saturating_add(1)),
+    )
+}
+
+struct ForwardingStageMetrics {
+    last_reported: Instant,
+    did_something: bool,
+
+    votes_received: usize,
+    votes_dropped_on_receive: usize,
+    votes_dropped_on_capacity: usize,
+    votes_dropped_on_data_budget: usize,
+    votes_forwarded: usize,
+
+    non_votes_received: usize,
+    non_votes_dropped_on_receive: usize,
+    non_votes_dropped_on_capacity: usize,
+    non_votes_dropped_on_data_budget: usize,
+    non_votes_forwarded: usize,
+
+    dropped_on_timeout: usize,
+}
+
+impl ForwardingStageMetrics {
+    fn maybe_report(&mut self) {
+        const REPORTING_INTERVAL: Duration = Duration::from_secs(1);
```

Comment:
regarding the choice of this `REPORTING_INTERVAL`, why it is 1 sec and not, lets say, per slot? I don't if we have aggreed on the metrics report time intervals and maybe it is out of scope of this PR, but it quite difficult to make sense of metrics which are aggregated using different time intervals. I would stick with per slot everywhere if there is no performance penalties associated with this.

Reply:
what slot though? this doesn't need or use the working bank to operate